solr/core/src/java/org/apache/solr/cluster/placement/impl/AttributeFetcherImpl.java

```diff
@@ -197,32 +148,78 @@ public AttributeValues fetchAttributes() {
       }
     }
 
-    return new AttributeValuesImpl(nodeToCoreCount,
-        nodeToDiskType,
-        nodeToFreeDisk,
-        nodeToTotalDisk,
-        nodeToHeapUsage,
-        nodeToSystemLoadAverage,
-        syspropSnitchToNodeToValue,
-        metricSnitchToNodeToValue);
+    for (Node node : nodeToReplicaInternalTags.keySet()) {
+      Set<String> tags = nodeToReplicaInternalTags.get(node);
+      Map<String, Map<String, List<Replica>>> infos = cloudManager.getNodeStateProvider().getReplicaInfo(node.getName(), tags);
+      infos.entrySet().stream()
+          .filter(entry -> requestedCollectionNamesMetrics.containsKey(entry.getKey()))
+          .forEach(entry -> {
+            CollectionMetricsBuilder collectionMetricsBuilder = collectionMetricsBuilders
+                .computeIfAbsent(entry.getKey(), c -> new CollectionMetricsBuilder());
+            entry.getValue().forEach((shardName, replicas) -> {
+              CollectionMetricsBuilder.ShardMetricsBuilder shardMetricsBuilder =
+                  collectionMetricsBuilder.getShardMetricsBuilders()
+                  .computeIfAbsent(shardName, s -> new CollectionMetricsBuilder.ShardMetricsBuilder());
+              replicas.forEach(replica -> {
+                CollectionMetricsBuilder.ReplicaMetricsBuilder replicaMetricsBuilder =
+                    shardMetricsBuilder.getReplicaMetricsBuilders()
+                    .computeIfAbsent(replica.getName(), n -> new CollectionMetricsBuilder.ReplicaMetricsBuilder());
+                replicaMetricsBuilder.setLeader(replica.isLeader());
+                if (replica.isLeader()) {
+                  shardMetricsBuilder.setLeaderMetrics(replicaMetricsBuilder);
+                }
+                Set<ReplicaMetric<?>> requestedMetrics = requestedCollectionNamesMetrics.get(replica.getCollection());
+                if (requestedMetrics == null) {
+                  throw new RuntimeException("impossible error");
```

Comment:
Likely an error log with the value of `replica.getCollection()` (or add the value to the exception text) could help understand how the impossible happens. Or just remove this check and let the NPE bubble up?

Reply:
I'll remove this check.